Add tests and various bugfixes

cuminandpaprika opened the pull request:
# Summary
This PR adds some basic unit tests and includes several bug fixes. Currently, the main sysl repo can't import the sysl-catalog master branch due to issues with the BlackFriday dependency.

## Proposed Changes:

- Add basic integration tests which use snapshots to verify generated output
- Add unit tests for query param, path param and return types
- Fix bugs in query param, path param and return type diagram generation

cuminandpaprika committed:
Set max retry to 100

Fixes an issue where requests to the PlantUML server fail when generating docs for large collections of sysl files

cuminandpaprika committed:
Rename files

Golang conventions generally follow snake_case or lowercase for filenames, e.g in the std lib, most files are lowercase.
Renaming diagram-creation to diagram improves the readability and means the test file isn't diagram-creation_test.go which sets off my OCD tendencies and just doesn't seem right.

There is an argument to be made that diagram.go is too similar to diagrams.go in catalogdiagrams, but I struggle to find a more appropriate name.

cuminandpaprika committed:
General tidiness improvements

Reorder functions for readability

CreatePathParamDataModel should be next to CreateParamDataModel

Add more detailed comments
Fix misspell in function name

joshcarp commented on go.mod:
```diff
@@ -15,11 +16,12 @@ require (
 	github.com/pkg/errors v0.9.1
 	github.com/radovskyb/watcher v1.0.7
-	github.com/russross/blackfriday v2.0.0+incompatible
+	github.com/russross/blackfriday/v2 v2.0.1
```
v2 doesn't work with go mods: https://github.com/russross/blackfriday/issues/500
Already talked with @cuminandpaprika about this

cuminandpaprika replied:
👍 Removed

joshcarp commented on go.mod:
```diff
@@ -15,11 +16,12 @@ require (
 	gopkg.in/alecthomas/kingpin.v2 v2.2.6
+	gotest.tools v2.2.0+incompatible
 )
-replace github.com/russross/blackfriday => github.com/russross/blackfriday v1.5.2
```
see above

cuminandpaprika replied:
Removed

joshcarp commented on pkg/catalog/diagram.go:
```diff
@@ -53,7 +52,7 @@ func (p *Generator) CreateMarkdown(t *template.Template, outputFileName string,
 	out := buf.Bytes()
 	if p.Format == "html" && !p.DisableCss {
-		raw := string(blackfriday.MarkdownCommon(out))
+		raw := string(blackfriday.Run(out))
```
See first comment

cuminandpaprika replied:
👍 Removed

joshcarp commented on pkg/catalog/diagram.go:
```diff
@@ -253,20 +288,12 @@ func (p *Generator) CreateFile(contents string, diagramType int, elems ...string
 		p.Log.Error(err)
 		return ""
 	}
-	newFileName := fileName
-	for i := 0; ; i++ {
-		if diagram, ok := targetMap[newFileName]; !ok || diagram == fileContents {
-			break
-		}
-		newFileName = strings.ReplaceAll(fileName, p.Ext, strconv.Itoa(i)+p.Ext)
-	}
-	fileName = newFileName
```
why did you remove this??

cuminandpaprika replied:
👍 Dropped that commit

joshcarp commented on pkg/catalog/diagram.go:
```diff
@@ -228,14 +265,12 @@ func (p *Generator) CreateTypeDiagram(app *sysl.Application, typeName string, t
 func CreateFileName(dir string, elems ...string) (string, string) {
 	absolutefileName := path.Join(append([]string{dir}, Map(elems, SanitiseOutputName)...)...)
 	relativefileName := strings.Replace(absolutefileName, dir+"/", "", 1)
-	absolutefileName = strings.ToLower(absolutefileName)
-	relativefileName = strings.ToLower(relativefileName)
```
strings.ToLower is needed so that there aren't filename clashes

cuminandpaprika replied:
👍 Dropped that commit

joshcarp commented on pkg/catalog/diagram.go:
```diff
@@ -279,48 +306,6 @@ func (p *Generator) GenerateDataModel(app *sysl.Application) string {
 }
 
-// CreateQueryParamDataModel returns a Query Parameter data model filename.
-func (p *Generator) CreateQueryParamDataModel(CurrentAppName string, param *sysl.Endpoint_RestParams_QueryParam) string {
```
??

cuminandpaprika replied:
Bungled rebase I think, fixed up in latest commits

joshcarp commented on pkg/catalog/template.go:
```diff
@@ -46,7 +46,7 @@ const NewPackageTemplate = `
 
 ### Request types
-{{if eq (len $e.Param) 0}}
+{{if and (eq (len $e.Param) 0) (eq $e.RestParams nil) }}
```
is this tested?
in go templating you can't equate a pointer type to nil: https://goplay.space/#QOWlt3VKeQr

`nil` is casted to `false`, same as any other zero type

cuminandpaprika replied:
Great catch, changed it to `with` instead of `if`

joshcarp commented on pkg/catalog/util.go:
```diff
@@ -135,18 +135,10 @@ func ModulePackageName(m *sysl.Module) string {
+func Map(vs []string, f func(string) string) []string {
 	vsm := make([]string, len(vs))
 	for i, v := range vs {
-		for j, f := range funcs {
-			var middle string
-			if j == 0 {
-				middle = f(v)
-				vsm[i] = middle
-			}
-			vsm[i] = f(middle)
-		}
-
```
??

cuminandpaprika replied:
Dropped this commit

joshcarp commented on pkg/catalog/diagram.go:
```diff
@@ -253,20 +288,12 @@ func (p *Generator) CreateFile(contents string, diagramType int, elems ...string
 		return fileContents
 	}
 	targetMap[fileName] = fileContents
-	return strings.Replace(fileName, p.CurrentDir+"/", "", 1)
```
this is still needed

cuminandpaprika replied:
Dropped this

joshcarp reviewed (changes requested):
I guess it's about time that this repo had some tests, but:
- You've reverted a commit (duplicate files) and removed a whole bunch of functionality that is needed. 
- You've added a bug into the templating (pointer equality to nil)
- The v2 change to blackfriday was very intentional, some users were experiencing problems with using `go get` with v2 (it's well documented on their repo)
- You've added a .snapshot folder with svgs, but plantuml svgs aren't deterministic, so diffs will always occur (is this expected, or are you only planning on checking that the files exist?, and if so, why not just have a set of filenames in the code?)

cuminandpaprika commented:
Thanks for the feedback, 

- yup, will fix up the reverted change
- ah, I didn’t know they were no deterministic, might need to use string finding instead. Do you know if what parts change in plantuml diagrams? Was it the ordering or the layout perhaps? I vaguely recall that happening with integration diagrams... 
- will fix the nil pointer issue, the error log would be super handy if you have it

Should I setup CI on this repo or are we intending to merge it into sysl in the future?

cuminandpaprika commented:
Also, is it feasible to replace blackfriday with goldmark instead so we can bypass all these go module issues? https://github.com/yuin/goldmark

It’s what hugo moved to from blackfriday

joshcarp commented:
- https://goplay.space/#QOWlt3VKeQr is the nil pointer issue; you don't need logs
- plantuml svgs are non deterministic, the plantuml code is (that's actually how sysl-catalog avoids geenerating duplicate diagrams; it stores everything in a map and if there's a match for the content it returns that filename)
- Which markdown generator is used doesn't matter

cuminandpaprika committed:
Add tests

Add high level integrated tests

Adds some high level tests for verifying file creation for the sysl-catalog repo

Add test target to makefile

joshcarp commented:
```
app:
    ep(req <: app2.this):
        return ok <: app2.this
app2:
    !type this:
        contents <: string
```

sysl-catalog --serve example.sysl:
![image](https://user-images.githubusercontent.com/32605850/82398853-2766a900-9a97-11ea-9de8-adc6be533516.png)

cuminandpaprika committed:
Fix path and query param diagrams

Adds unit tests for path and query param diagram generation

Go mod tidy

fixup! Fix path and query param diagrams

Fix bug in type ref to other apps

Fix broken diagrams for sequences

cuminandpaprika commented:
> ```
> app:
>     ep(req <: app2.this):
>         return ok <: app2.this
> app2:
>     !type this:
>         contents <: string
> ```
> 
> sysl-catalog --serve example.sysl:
> ![image](https://user-images.githubusercontent.com/32605850/82398853-2766a900-9a97-11ea-9de8-adc6be533516.png)

Thanks, fixed up! along with a bunch of other issues with diagram generation!
